lib/collection/src/shards/shard_holder.rs

```diff
@@ -165,6 +165,57 @@ impl ShardHolder {
         Ok(())
     }
 
+    pub async fn abort_resharding(
+        &mut self,
+        shard_id: ShardId,
+        peer_id: PeerId,
+        shard_key: Option<ShardKey>,
+    ) -> Result<(), CollectionError> {
+        if let Some(ring) = self.rings.get_mut(&shard_key) {
+            log::debug!("removing peer {peer_id} from {shard_key:?} hashring");
+            ring.remove_resharding(shard_id);
+        } else {
+            log::warn!(
+                "aborting resharding of shard {shard_id} ({peer_id}/{shard_key:?}), \
+                 but {shard_key:?} hashring does not exist"
+            );
+        }
+
+        if let Some(shard) = self.get_shard(&shard_id) {
+            match shard.peer_state(&peer_id) {
+                Some(ReplicaState::Resharding) => {
+                    log::debug!("removing peer {peer_id} from {shard_id} replica set");
+                    shard.remove_peer(peer_id).await?;
+                }
+
+                Some(state) => {
+                    return Err(CollectionError::bad_request(format!(
+                        "peer {peer_id} is in {state:?} state"
+                    )));
+                }
```

Comment:
We probably also want to allow dead state?

Reply:
Hmmm, good point. I'd expect `Dead` peers to automatically stop resharding, but one of the points of `Abort` is to be able to resolve broken state.

I'll see if there's a **neat/safe** way to add this. You would not want to accidentally yeet *any* random `Dead` peer using this.

Reply:
If that shard ID is in the resharding state, I think it should be fine to yeet it in dead state.

Reply:
Done. Might have overcomplicated this check a bit, but we store resharding state in a bunch of places, so I had to come up with something.